README.md
# EthereumAbi

This is part of implementation of [ruby-ethereum](https://github.com/cryptape/ruby-ethereum)
that related to calculation of ABI Solidity contract calls for the EVM.

ABI stands for application binary interface.

In general, an ABI is the interface between two program modules, one of which is often at the level of machine code.
The interface is the de facto method for encoding/decoding data into/out of the machine code.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'ethereum_abi'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install ethereum_abi

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/TemaMix/ethereum_abi. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/TemaMix/ethereum_abi/blob/main/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the EthereumAbi project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/TemaMix/ethereum_abi/blob/main/CODE_OF_CONDUCT.md).


CHANGELOG.md
## [Released]

## [0.1.1] - 2022-09-23
- Fix path to EthereumAbi::Constant

## [0.1.0] - 2022-09-21

- Initial release
- Copy ABI from https://github.com/cryptape/ruby-ethereum


spec/spec_helper.rb
# -*- encoding : ascii-8bit -*-
# frozen_string_literal: true

require "ethereum_abi"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


test/test_helper.rb
# -*- encoding : ascii-8bit -*-
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require "ethereum_abi"

require "minitest/autorun"


spec/ethereum_abi_spec.rb
# -*- encoding : ascii-8bit -*-
# frozen_string_literal: true

require "spec_helper"

RSpec.describe EthereumAbi do
  shared_examples "Encode ABI for type" do |type, values|
    it "encodes #{type} value" do
      values.each do |value|
        expect(
          EthereumAbi.encode_primitive_type(EthereumAbi::Type.parse(type), value)
        ).to eq(EthereumAbi.encode_abi([type], [value]))
      end
    end
  end

  describe ".encode_abi" do
    include_examples("Encode ABI for type", "int8", [1, -1, 127, -128])
    include_examples("Encode ABI for type", "int32", [1, -1, 127, -128, 2**31 - 1, -2**31])
    include_examples("Encode ABI for type", "int256", [1, -1, 127, -128, 2**31 - 1, -2**31, 2**255 - 1, -2**255])
  end

  describe ".encode_type" do
  end

  shared_examples "Encode ABI for primitive type" do |type, values|
    it "encodes #{type} value" do
      values.each do |value|
        expect(
          EthereumAbi.encode_primitive_type(EthereumAbi::Type.parse(type), value)
        ).to eq(EthereumAbi.encode_abi([type], [value]))
      end
    end
  end

  describe ".encode_primitive_type" do
    it "encodes bool type" do
      type = EthereumAbi::Type.parse "bool"
      expect(EthereumAbi.encode_primitive_type(type, true)).to eq(EthereumAbi::Utils.zpad_int(1))
      expect(EthereumAbi.encode_primitive_type(type, false)).to eq(EthereumAbi::Utils.zpad_int(0))
    end

    it "encodes uint8 type" do
      type = EthereumAbi::Type.parse "uint8"
      expect(EthereumAbi.encode_primitive_type(type, 255)).to eq(EthereumAbi::Utils.zpad_int(255))
      expect { EthereumAbi.encode_primitive_type(type, 256) }.to raise_error(EthereumAbi::ValueOutOfBounds)
    end

    it "encodes ufixed128x128 type" do
      type = EthereumAbi::Type.parse "ufixed128x128"
      # expect(EthereumAbi.encode_primitive_type(type, 0)).to eq("\x00"*32)
      # expect(EthereumAbi.encode_primitive_type(type, 1.125)).to eq("\x00"*15 + "\x01\x20" + "\x00"*15)
      expect(EthereumAbi.encode_primitive_type(type, 2**127 - 1)).to eq("\x7F#{"\xff" * 15}#{"\x00" * 16}")
    end

    it "encodes fixed128x128 type" do
      type = EthereumAbi::Type.parse "fixed128x128"
      expect(EthereumAbi.encode_primitive_type(type, -1)).to eq("\xff" * 16 + "\x00" * 16)
      expect(EthereumAbi.encode_primitive_type(type, -2**127)).to eq("\x80#{"\x00" * 31}")
      expect(EthereumAbi.encode_primitive_type(type, 2**127 - 1)).to eq("\x7F#{"\xff" * 15}#{"\x00" * 16}")
      expect(EthereumAbi.encode_primitive_type(type, 1.125)).to eq("#{EthereumAbi::Utils.zpad_int(1, 16)}\x20#{"\x00" * 15}")
      expect(EthereumAbi.encode_primitive_type(type, -1.125)).to eq("#{"\xff" * 15}\xfe\xe0#{"\x00" * 15}")
      expect { EthereumAbi.encode_primitive_type(type, -2**127 - 1) }.to raise_error(EthereumAbi::ValueOutOfBounds)
      expect { EthereumAbi.encode_primitive_type(type, 2**127) }.to raise_error(EthereumAbi::ValueOutOfBounds)
    end

    it "encodes byte type" do
      type = EthereumAbi::Type.parse "bytes"
      expect(
        EthereumAbi.encode_primitive_type(type, "\x01\x02\x03")
      ).to eq("#{EthereumAbi::Utils.zpad_int(3)}\x01\x02\x03#{"\x00" * 29}")
    end

    it "encodes fixed128x128 type" do
      type = EthereumAbi::Type.parse "bytes8"
      expect(
        EthereumAbi.encode_primitive_type(type, "\x01\x02\x03")
      ).to eq("\x01\x02\x03#{"\x00" * 29}")
    end

    it "encodes hash32 type" do
      type = EthereumAbi::Type.parse "hash32"
      expect(
        EthereumAbi.encode_primitive_type(type, "\xff" * 32)
      ).to eq("\xff" * 32)
      expect(
        EthereumAbi.encode_primitive_type(type, "ff" * 32)
      ).to eq("\xff" * 32)
    end

    it "encodes address type" do
      type = EthereumAbi::Type.parse "address"
      expect(
        EthereumAbi.encode_primitive_type(type, "\xff" * 20)
      ).to eq(EthereumAbi::Utils.zpad("\xff" * 20, 32))
      expect(
        EthereumAbi.encode_primitive_type(type, "ff" * 20)
      ).to eq(EthereumAbi::Utils.zpad("\xff" * 20, 32))
      expect(
        EthereumAbi.encode_primitive_type(type, "0x#{"ff" * 20}")
      ).to eq(EthereumAbi::Utils.zpad("\xff" * 20, 32))
    end
  end

  shared_examples "Decode ABI for type" do |type, values|
    it "encodes #{type} value" do
      values.each do |value|
        expect(
          EthereumAbi.decode_abi([type], EthereumAbi.encode_abi([type], [value]))[0]
        ).to eq(value)
      end
    end
  end

  describe ".decode_abi" do
    include_examples("Decode ABI for type", "int8", [1, -1, 127, -128])
    include_examples("Decode ABI for type", "int32", [1, -1, 127, -128, 2**31 - 1, -2**31])
    include_examples("Decode ABI for type", "int256", [1, -1, 127, -128, 2**31 - 1, -2**31, 2**255 - 1, -2**255])
  end

  describe ".decode_type" do
  end

  describe ".decode_primitive_type" do
    it "decodes address type" do
      type = EthereumAbi::Type.parse "address"
      expect(
        EthereumAbi.decode_primitive_type(
          type,
          EthereumAbi.encode_primitive_type(type, "0x#{"ff" * 20}")
        )
      ).to eq("ff" * 20)
    end

    it "decodes bytes type" do
      type = EthereumAbi::Type.parse "bytes"
      expect(
        EthereumAbi.decode_primitive_type(
          type,
          EthereumAbi.encode_primitive_type(type, "\x01\x02\x03")
        )
      ).to eq("\x01\x02\x03")
    end

    it "decodes bytes8 type" do
      type = EthereumAbi::Type.parse "bytes8"
      expect(
        EthereumAbi.decode_primitive_type(
          type,
          EthereumAbi.encode_primitive_type(type, "\x01\x02\x03")
        )
      ).to eq("\x01\x02\x03#{"\x00" * 5}")
    end

    it "decodes hash20 type" do
      type = EthereumAbi::Type.parse "hash20"
      expect(
        EthereumAbi.decode_primitive_type(
          type,
          EthereumAbi.encode_primitive_type(type, "ff" * 20)
        )
      ).to eq("\xff" * 20)
    end

    it "decodes uint8 type" do
      type = EthereumAbi::Type.parse "uint8"
      expect(
        EthereumAbi.decode_primitive_type(
          type,
          EthereumAbi.encode_primitive_type(type, 0)
        )
      ).to eq(0)
      expect(
        EthereumAbi.decode_primitive_type(
          type,
          EthereumAbi.encode_primitive_type(type, 255)
        )
      ).to eq(255)
    end

    it "decodes int8 type" do
      type = EthereumAbi::Type.parse "int8"
      expect(
        EthereumAbi.decode_primitive_type(
          type,
          EthereumAbi.encode_primitive_type(type, -128)
        )
      ).to eq(-128)
      expect(
        EthereumAbi.decode_primitive_type(
          type,
          EthereumAbi.encode_primitive_type(type, 127)
        )
      ).to eq(127)
    end

    it "decodes ufixed128x128 type" do
      type = EthereumAbi::Type.parse "ufixed128x128"
      expect(
        EthereumAbi.decode_primitive_type(
          type,
          EthereumAbi.encode_primitive_type(type, 0)
        )
      ).to eq(0)
      expect(
        EthereumAbi.decode_primitive_type(
          type,
          EthereumAbi.encode_primitive_type(type, 125.125)
        )
      ).to eq(125.125)
      expect(
        EthereumAbi.decode_primitive_type(
          type,
          EthereumAbi.encode_primitive_type(type, 2**128 - 1)
        )
      ).to eq((2**128 - 1).to_f)
    end

    it "decodes fixed128x128 type" do
      type = EthereumAbi::Type.parse "fixed128x128"
      expect(
        EthereumAbi.decode_primitive_type(
          type,
          EthereumAbi.encode_primitive_type(type, 1)
        )
      ).to eq(1)
      expect(
        EthereumAbi.decode_primitive_type(
          type,
          EthereumAbi.encode_primitive_type(type, -1)
        )
      ).to eq(-1)
      expect(
        EthereumAbi.decode_primitive_type(
          type,
          EthereumAbi.encode_primitive_type(type, 125.125)
        )
      ).to eq(125.125)
      expect(
        EthereumAbi.decode_primitive_type(
          type,
          EthereumAbi.encode_primitive_type(type, -125.125)
        )
      ).to eq(-125.125)
      expect(
        EthereumAbi.decode_primitive_type(
          type,
          EthereumAbi.encode_primitive_type(type, 2**127 - 1)
        )
      ).to eq((2**127-1).to_f)
      expect(
        EthereumAbi.decode_primitive_type(
          type,
          EthereumAbi.encode_primitive_type(type, -2**127)
        )
      ).to eq(-2**127)
    end

    it "decodes bool type" do
      type = EthereumAbi::Type.parse "bool"
      expect(
        EthereumAbi.decode_primitive_type(
          type,
          EthereumAbi.encode_primitive_type(type, true)
        )
      ).to eq(true)
      expect(
        EthereumAbi.decode_primitive_type(
          type,
          EthereumAbi.encode_primitive_type(type, false)
        )
      ).to eq(false)
    end
  end
end


lib/ethereum_abi.rb
# -*- encoding : ascii-8bit -*-
# frozen_string_literal: true

require "json"
require "rlp"

require_relative "ethereum_abi/version"
require_relative "ethereum_abi/utils"
require_relative "ethereum_abi/constant"
require_relative "ethereum_abi/type"

module EthereumAbi
  class Error < StandardError; end
  class ValueError < StandardError; end

  extend self

  include EthereumAbi::Constant

  class EncodingError < StandardError; end
  class DecodingError < StandardError; end
  class ValueOutOfBounds < StandardError; end

  ##
  # Encodes multiple arguments using the head/tail mechanism.
  #
  def encode_abi(types, args)
    parsed_types = types.map { |t| Type.parse(t) }

    head_size = (0...args.size)
                .map { |i| parsed_types[i].size || 32 }
                .reduce(0, &:+)

    head = ""
    tail = ""
    args.each_with_index do |arg, i|
      if parsed_types[i].dynamic?
        head += encode_type(Type.size_type, head_size + tail.size)
        tail += encode_type(parsed_types[i], arg)
      else
        head += encode_type(parsed_types[i], arg)
      end
    end

    "#{head}#{tail}"
  end
  alias encode encode_abi

  ##
  # Encodes a single value (static or dynamic).
  #
  # @param type [EthereumAbi::Type] value type
  # @param arg [Object] value
  #
  # @return [String] encoded bytes
  #
  def encode_type(type, arg)
    if %w[string bytes].include?(type.base) && type.sub.empty?
      raise ArgumentError, "arg must be a string" unless arg.instance_of?(String)

      size = encode_type Type.size_type, arg.size
      padding = BYTE_ZERO * (Utils.ceil32(arg.size) - arg.size)

      "#{size}#{arg}#{padding}"
    elsif type.dynamic?
      raise ArgumentError, "arg must be an array" unless arg.instance_of?(Array)

      head = ""
      tail = ""
      if type.dims.last.zero?
        head += encode_type(Type.size_type, arg.size)
      else
        unless arg.size == type.dims.last
          raise ArgumentError,
                "Wrong array size: found #{arg.size}, expecting #{type.dims.last}"
        end
      end

      sub_type = type.subtype
      sub_size = type.subtype.size
      arg.size.times do |i|
        if sub_size.nil?
          head += encode_type(Type.size_type, 32 * arg.size + tail.size)
          tail += encode_type(sub_type, arg[i])
        else
          head += encode_type(sub_type, arg[i])
        end
      end

      "#{head}#{tail}"
    elsif type.dims.empty? # static type
      encode_primitive_type type, arg
    else
      arg.map { |x| encode_type(type.subtype, x) }.join
    end
  end

  def encode_primitive_type(type, arg)
    case type.base
    when "uint"
      real_size = type.sub.to_i
      i = get_uint arg

      raise ValueOutOfBounds, arg unless i >= 0 && i < 2**real_size

      Utils.zpad_int i
    when "bool"
      raise ArgumentError, "arg is not bool: #{arg}" unless arg.instance_of?(TrueClass) || arg.instance_of?(FalseClass)

      Utils.zpad_int(arg ? 1 : 0)
    when "int"
      real_size = type.sub.to_i
      i = get_int arg

      raise ValueOutOfBounds, arg unless i >= -2**(real_size - 1) && i < 2**(real_size - 1)

      Utils.zpad_int(i % 2**type.sub.to_i)
    when "ureal", "ufixed"
      high, low = type.sub.split("x").map(&:to_i)

      raise ValueOutOfBounds, arg unless arg >= 0 && arg < 2**high

      Utils.zpad_int((arg * 2**low).to_i)
    when "real", "fixed"
      high, low = type.sub.split("x").map(&:to_i)

      raise ValueOutOfBounds, arg unless arg >= -2**(high - 1) && arg < 2**(high - 1)

      i = (arg * 2**low).to_i
      Utils.zpad_int(i % 2**(high + low))
    when "string", "bytes"
      raise EncodingError, "Expecting string: #{arg}" unless arg.instance_of?(String)

      if type.sub.empty? # variable length type
        size = Utils.zpad_int arg.size
        padding = BYTE_ZERO * (Utils.ceil32(arg.size) - arg.size)
        "#{size}#{arg}#{padding}"
      else # fixed length type
        raise ValueOutOfBounds, arg unless arg.size <= type.sub.to_i

        padding = BYTE_ZERO * (32 - arg.size)
        "#{arg}#{padding}"
      end
    when "hash"
      size = type.sub.to_i
      raise EncodingError, "too long: #{arg}" unless size.positive? && size <= 32

      if arg.is_a?(Integer)
        Utils.zpad_int(arg)
      elsif arg.size == size
        Utils.zpad arg, 32
      elsif arg.size == size * 2
        Utils.zpad_hex arg
      else
        raise EncodingError, "Could not parse hash: #{arg}"
      end
    when "address"
      if arg.is_a?(Integer)
        Utils.zpad_int arg
      elsif arg.size == 20
        Utils.zpad arg, 32
      elsif arg.size == 40
        Utils.zpad_hex arg
      elsif arg.size == 42 && arg[0, 2] == "0x"
        Utils.zpad_hex arg[2..-1]
      else
        raise EncodingError, "Could not parse address: #{arg}"
      end
    else
      raise EncodingError, "Unhandled type: #{type.base} #{type.sub}"
    end
  end

  ##
  # Decodes multiple arguments using the head/tail mechanism.
  #
  def decode_abi(types, data)
    parsed_types = types.map { |t| Type.parse(t) }

    outputs = [nil] * types.size
    start_positions = [nil] * types.size + [data.size]

    pos = 0
    parsed_types.each_with_index do |t, i|
      # If a type is static, grab the data directly, otherwise record its
      # start position
      if t.dynamic?
        start_positions[i] = Utils.big_endian_to_int(data[pos, 32])

        j = i - 1
        while j >= 0 && start_positions[j].nil?
          start_positions[j] = start_positions[i]
          j -= 1
        end

        pos += 32
      else
        outputs[i] = data[pos, t.size]
        pos += t.size
      end
    end

    # We add a start position equal to the length of the entire data for
    # convenience.
    j = types.size - 1
    while j >= 0 && start_positions[j].nil?
      start_positions[j] = start_positions[types.size]
      j -= 1
    end

    raise DecodingError, "Not enough data for head" unless pos <= data.size

    parsed_types.each_with_index do |t, i|
      if t.dynamic?
        offset, next_offset = start_positions[i, 2]
        outputs[i] = data[offset...next_offset]
      end
    end

    parsed_types.zip(outputs).map { |(type, out)| decode_type(type, out) }
  end
  alias decode decode_abi

  def decode_type(type, arg)
    if %w[string bytes].include?(type.base) && type.sub.empty?
      l = Utils.big_endian_to_int arg[0, 32]
      data = arg[32..-1]

      raise DecodingError, "Wrong data size for string/bytes object" unless data.size == Utils.ceil32(l)

      data[0, l]
    elsif type.dynamic?
      l = Utils.big_endian_to_int arg[0, 32]
      subtype = type.subtype

      if subtype.dynamic?
        raise DecodingError, "Not enough data for head" unless arg.size >= 32 + 32 * l

        start_positions = (1..l).map { |i| Utils.big_endian_to_int arg[32 * i, 32] }
        start_positions.push arg.size

        outputs = (0...l).map { |i| arg[start_positions[i]...start_positions[i + 1]] }

        outputs.map { |out| decode_type(subtype, out) }
      else
        (0...l).map { |i| decode_type(subtype, arg[32 + subtype.size * i, subtype.size]) }
      end
    elsif !type.dims.empty? # static-sized arrays
      l = type.dims.last[0]
      subtype = type.subtype

      (0...l).map { |i| decode_type(subtype, arg[subtype.size * i, subtype.size]) }
    else
      decode_primitive_type type, arg
    end
  end

  def decode_primitive_type(type, data)
    case type.base
    when "address"
      Utils.encode_hex data[12..-1]
    when "string", "bytes"
      if type.sub.empty? # dynamic
        size = Utils.big_endian_to_int data[0, 32]
        data[32..-1][0, size]
      else # fixed
        data[0, type.sub.to_i]
      end
    when "hash"
      data[(32 - type.sub.to_i), type.sub.to_i]
    when "uint"
      Utils.big_endian_to_int data
    when "int"
      u = Utils.big_endian_to_int data
      u >= 2**(type.sub.to_i - 1) ? (u - 2**type.sub.to_i) : u
    when "ureal", "ufixed"
      _high, low = type.sub.split("x").map(&:to_i)
      Utils.big_endian_to_int(data) * 1.0 / 2**low
    when "real", "fixed"
      high, low = type.sub.split("x").map(&:to_i)
      u = Utils.big_endian_to_int data
      i = u >= 2**(high + low - 1) ? (u - 2**(high + low)) : u
      i * 1.0 / 2**low
    when "bool"
      data[-1] == BYTE_ONE
    else
      raise DecodingError, "Unknown primitive type: #{type.base}"
    end
  end

  private

  def get_uint(n)
    case n
    when Integer
      raise EncodingError, "Number out of range: #{n}" if n > UINT_MAX || n < UINT_MIN

      n
    when String
      if n.size == 40
        Utils.big_endian_to_int Utils.decode_hex(n)
      elsif n.size <= 32
        Utils.big_endian_to_int n
      else
        raise EncodingError, "String too long: #{n}"
      end
    when true
      1
    when false, nil
      0
    else
      raise EncodingError, "Cannot decode uint: #{n}"
    end
  end

  def get_int(n)
    case n
    when Integer
      raise EncodingError, "Number out of range: #{n}" if n > INT_MAX || n < INT_MIN

      n
    when String
      if n.size == 40
        i = Utils.big_endian_to_int Utils.decode_hex(n)
      elsif n.size <= 32
        i = Utils.big_endian_to_int n
      else
        raise EncodingError, "String too long: #{n}"
      end
      i > INT_MAX ? (i - TT256) : i
    when true
      1
    when false, nil
      0
    else
      raise EncodingError, "Cannot decode int: #{n}"
    end
  end
end


lib/ethereum_abi/utils.rb
# -*- encoding : ascii-8bit -*-
# frozen_string_literal: true

require_relative "constant"

module EthereumAbi
  module Utils
    extend self

    include EthereumAbi::Constant

    def ceil32(x)
      (x % 32).zero? ? x : (x + 32 - x % 32)
    end

    def encode_hex(b)
      ::RLP::Utils.encode_hex b
    end

    def decode_hex(s)
      ::RLP::Utils.decode_hex s
    end

    def big_endian_to_int(s)
      ::RLP::Sedes.big_endian_int.deserialize s.sub(/\A(\x00)+/, "")
    end

    def int_to_big_endian(n)
      ::RLP::Sedes.big_endian_int.serialize n
    end

    def lpad(x, symbol, l)
      return x if x.size >= l

      symbol * (l - x.size) + x
    end

    def rpad(x, symbol, l)
      return x if x.size >= l

      x + symbol * (l - x.size)
    end

    def zpad(x, l)
      lpad x, BYTE_ZERO, l
    end

    def zunpad(x)
      x.sub(/\A\x00+/, "")
    end

    def zpad_int(n, l = 32)
      zpad encode_int(n), l
    end

    def zpad_hex(s, l = 32)
      zpad decode_hex(s), l
    end

    def encode_int(n)
      raise ArgumentError, "Integer invalid or out of range: #{n}" unless n.is_a?(Integer) && n >= 0 && n <= UINT_MAX

      int_to_big_endian n
    end

    def decode_int(v)
      if v.size.positive? && (v[0] == Constant::BYTE_ZERO || v[0].zero?)
        raise ArgumentError,
              "No leading zero bytes allowed for integers"
      end

      big_endian_to_int v
    end
  end
end


lib/ethereum_abi/contract_translator.rb
# -*- encoding : ascii-8bit -*-
# frozen_string_literal: true

module EthereumAbi
  class ContractTranslator
    def initialize(contract_interface)
      contract_interface = JSON.parse contract_interface if contract_interface.instance_of?(String)

      @contract = {
        constructor_data: nil,
        function_data: {},
        event_data: {}
      }

      contract_interface.each do |desc|
        encode_types = desc["inputs"].map { |e| e["type"] }
        signature = desc["inputs"].map { |e| [e["type"], e["name"]] }

        # type can be omitted, defaulting to function
        type = desc["type"] || "function"
        case type
        when "function"
          name = basename desc["name"]
          decode_types = desc["outputs"].map { |e| e["type"] }
          @contract[:function_data][name] = {
            prefix: method_id(name, encode_types),
            encode_types: encode_types,
            decode_types: decode_types,
            is_constant: desc.fetch("constant", false),
            signature: signature
          }
        when "event"
          name = basename desc["name"]
          indexed = desc["inputs"].map { |e| e["indexed"] }
          names = desc["inputs"].map { |e| e["name"] }
          @contract[:event_data][event_id(name, encode_types)] = {
            types: encode_types,
            name: name,
            names: names,
            indexed: indexed,
            anonymous: desc.fetch("anonymous", false)
          }
        when "constructor"
          raise ValueError, "Only one constructor is supported." if @contract[:constructor_data]

          @contract[:constructor_data] = {
            encode_types: encode_types,
            signature: signature
          }
        else
          raise ValueError, "Unknown interface type: #{type}"
        end
      end
    end

    ##
    # Return the encoded function call.
    #
    # @param name [String] One of the existing functions described in the
    #   contract interface.
    # @param args [Array[Object]] The function arguments that will be encoded
    #   and used in the contract execution in the vm.
    #
    # @return [String] The encoded function name and arguments so that it can
    #   be used with the evm to execute a function call, the binary string
    #   follows the Ethereum Contract ABI.
    #
    def encode(name, args)
      raise ValueError, "Unknown function #{name}" unless function_data.include?(name)

      desc = function_data[name]
      func_id = Utils.zpad(Utils.encode_int(desc[:prefix]), 4)
      calldata = EthereumAbi.encode_abi desc[:encode_types], args

      "#{func_id}#{calldata}"
    end

    ##
    # Return the encoded constructor call.
    #
    def encode_constructor_arguments(args)
      raise ValueError, "The contract interface didn't have a constructor" unless constructor_data

      EthereumAbi.encode_abi constructor_data[:encode_types], args
    end

    def decode(name, data)
      desc = function_data[name]
      EthereumAbi.decode_abi desc[:decode_types], data
    end

    def constructor_data
      @contract[:constructor_data]
    end

    def function_data
      @contract[:function_data]
    end

    def event_data
      @contract[:event_data]
    end

    def function(name)
      function_data[name]
    end

    def event(name, encode_types)
      event_data[event_id(name, encode_types)]
    end

    def listen(log, noprint: false)
      return if log.topics.size.zero? || !event_data.key?(log.topics[0])

      data = event_data[log.topics[0]]
      types = data[:types]
      name = data[:name]
      names = data[:names]
      indexed = data[:indexed]
      indexed_types = types.zip(indexed).select { |(_t, i)| i.true? }.map(&:first)
      unindexed_types = types.zip(indexed).select { |(_t, i)| i.false? }.map(&:first)

      deserialized_args = EthereumAbi.decode_abi unindexed_types, log.data

      o = {}
      c1 = 0
      c2 = 0
      names.each_with_index do |n, i|
        if indexed[i].true?
          topic_bytes = Utils.zpad_int log.topics[c1 + 1]
          o[n] = EthereumAbi.decode_primitive_type ABI::Type.parse(indexed_types[c1]), topic_bytes
          c1 += 1
        else
          o[n] = deserialized_args[c2]
          c2 += 1
        end
      end

      o["_event_type"] = name
      p o unless noprint

      o
    end

    def method_id(name, encode_types)
      Utils.big_endian_to_int Utils.keccak256(get_sig(name, encode_types))[0, 4]
    end

    def event_id(name, encode_types)
      Utils.big_endian_to_int Utils.keccak256(get_sig(name, encode_types))
    end

    private

    def logger
      @logger ||= Logger.new "eth.abi.contract_translator"
    end

    def get_sig(name, encode_types)
      "#{name}(#{encode_types.map { |x| canonical_name(x) }.join(",")})"
    end

    def canonical_name(x)
      case x
      when /\A(uint|int)(\[.*\])?\z/
        "#{Regexp.last_match(1)}256#{Regexp.last_match(2)}"
      when /\A(real|ureal|fixed|ufixed)(\[.*\])?\z/
        "#{Regexp.last_match(1)}128x128#{Regexp.last_match(2)}"
      else
        x
      end
    end

    def basename(n)
      i = n.index "("
      i ? n[0, i] : n
    end
  end
end


lib/ethereum_abi/type.rb
# -*- encoding : ascii-8bit -*-
# frozen_string_literal: true

module EthereumAbi
  class Type
    class ParseError < StandardError; end

    class << self
      ##
      # Crazy regexp to seperate out base type component (eg. uint), size (eg.
      # 256, 128x128, nil), array component (eg. [], [45], nil)
      #
      def parse(type)
        _, base, sub, dimension = /([a-z]*)([0-9]*x?[0-9]*)((\[[0-9]*\])*)/.match(type).to_a

        dims = dimension.scan(/\[[0-9]*\]/)
        raise ParseError, "Unknown characters found in array declaration" if dims.join != dimension

        case base
        when "string"
          raise ParseError, "String type must have no suffix or numerical suffix" unless sub.empty?
        when "bytes"
          raise ParseError, "Maximum 32 bytes for fixed-length string or bytes" unless sub.empty? || sub.to_i <= 32
        when "uint", "int"
          raise ParseError, "Integer type must have numerical suffix" unless sub =~ /\A[0-9]+\z/

          size = sub.to_i
          raise ParseError, "Integer size out of bounds" unless size >= 8 && size <= 256
          raise ParseError, "Integer size must be multiple of 8" unless (size % 8).zero?
        when "ureal", "real", "fixed", "ufixed"
          unless sub =~ /\A[0-9]+x[0-9]+\z/
            raise ParseError,
                  "Real type must have suffix of form <high>x<low>, e.g. 128x128"
          end

          high, low = sub.split("x").map(&:to_i)
          total = high + low

          raise ParseError, "Real size out of bounds (max 32 bytes)" unless total >= 8 && total <= 256
          raise ParseError, "Real high/low sizes must be multiples of 8" unless (high % 8).zero? && (low % 8).zero?
        when "hash"
          raise ParseError, "Hash type must have numerical suffix" unless sub =~ /\A[0-9]+\z/
        when "address"
          raise ParseError, "Address cannot have suffix" unless sub.empty?
        when "bool"
          raise ParseError, "Bool cannot have suffix" unless sub.empty?
        else
          raise ParseError, "Unrecognized type base: #{base}"
        end

        new(base, sub, dims.map { |x| x[1...-1].to_i })
      end

      def size_type
        @size_type ||= new("uint", 256, [])
      end
    end

    attr :base, :sub, :dims

    ##
    # @param base [String] base name of type, e.g. uint for uint256[4]
    # @param sub  [String] subscript of type, e.g. 256 for uint256[4]
    # @param dims [Array[Integer]] dimensions of array type, e.g. [1,2,0]
    #   for uint256[1][2][], [] for non-array type
    #
    def initialize(base, sub, dims)
      @base = base
      @sub  = sub
      @dims = dims
    end

    def ==(other)
      base == other.base &&
        sub == other.sub &&
        dims == other.dims
    end

    ##
    # Get the static size of a type, or nil if dynamic.
    #
    # @return [Integer, NilClass]  size of static type, or nil for dynamic
    #   type
    #
    def size
      @size ||= if dims.empty?
                  if %w[string bytes].include?(base) && sub.empty?
                    nil
                  else
                    32
                  end
                elsif dims.last.zero?
                  nil # 0 for dynamic array []
                else
                  subtype.dynamic? ? nil : dims.last * subtype.size
                end
    end

    def dynamic?
      size.nil?
    end

    ##
    # Type with one dimension lesser.
    #
    # @example
    #   Type.parse("uint256[2][]").subtype # => Type.new('uint', 256, [2])
    #
    # @return [EthereumAbi::Type]
    #
    def subtype
      @subtype ||= self.class.new(base, sub, dims[0...-1])
    end
  end
end
